src/ucp/core/ucp_mm.c

```diff
@@ -106,6 +106,32 @@ static ucs_status_t ucp_memh_reg_mds(ucp_context_h context, ucp_mem_h memh,
     return UCS_OK;
 }
 
+ucs_status_t ucp_memory_type_detect_mds(ucp_context_h context, void *addr, size_t length,
+                                        ucp_mem_type_t *mem_type)
+{
+    unsigned md_index;
+
+    mem_type->md_map = 0;
+    mem_type->id = UCT_MD_MEM_TYPE_HOST;
+
+    if (!context->num_mem_type_mds) {
```

Comment:
since you already count "num_mem_type_mds", why not keep them all in an array inside context, so would not have to check irrelevant md's on the fast path?